### PR TITLE
Add DMRE to Data Matrix

### PR DESCRIPTION
**Pull request by gitlost**
Adds DMRE (ISO/IEC 21471:2020) to Data Matrix.

The PR includes a number of changes to `DataMatrix::DetectOld()` which seemed to have some regressions compared to the Java upstream. They've been checked by hacking `DataMatrix::Detect()` to just call it instead and now fail for the same test cases but obviously could do with some close inspection.

The test samples were generated by ZInt plus 2 jpeg ones from the [EuroData Council](https://e-d-c.info/en/projects/dmre-en.html) site (after being processed a bit).

### CODE REVIEW

**Review by axxel — CHANGES_REQUESTED** [review]
Thanks a lot. I wonder though, why you mixed the support for DMRE and some seemingly unrelated fixes/improvements to the `DecodeOld` function in one commit/PR. Could you please separate that, ideally in two PRs?

File: core/src/datamatrix/DMDecoder.cpp
Comment (gitlost) [reply core/src/datamatrix/DMDecoder.cpp]: Done.

File: core/src/datamatrix/DMBitLayout.cpp
Comment (gitlost) [reply core/src/datamatrix/DMBitLayout.cpp:84]: Yes, it's an addition to the ISO 16022:2006 F.1 Symbol placement character `module()` routine in the ISO 21471:2020 equivalent E.1 Symbol placement character `module()` routine.

File: test/blackbox/BlackboxTestRunner.cpp
Comment (gitlost) [reply test/blackbox/BlackboxTestRunner.cpp]: Done to `datamatrix-4`.

**Comment by axxel** [comment]
That was fast! :)

I'm currently waiting for some feedback from the owner of the EuroData Council site where you took the two images from to see if he has objections to that usage.

**Comment by axxel** [comment]
I did not get any feedback regarding the 2 images but don't want to let you wait any longer.  If you'd like to send the DetectOld improvements as well, please go ahead. Thanks.

**Comment by gitlost** [comment]
Thanks, re DetectOld changes, I'll leave them for now and press on with the other PRs I have. They're by far the dodgiest changes I'm proposing and need some way of being tested apart from hacking the code - e.g. by a test-only `useOld`/`alternate`/`upstream` option being added?

Re image permission, the main contact on the EuroData site is Heinrich Oehlmann, who is the father of Harald Oehlmann (@oehhar), a principal maintainer of and contributor to Zint, so pinging him here as maybe he can help...

**Pull request merged**
